File: cn/java-springmvc/src/main/java/com/example/controller/CategoryController.java
package com.example.controller;

import org.springframework.web.bind.annotation.*;
import com.example.model.Category;
import com.example.dto.Result;
import com.example.dto.PageResult;
import jakarta.validation.Valid;
import java.util.List;

@RestController
@RequestMapping("/api/categories")
public class CategoryController {
    
    @GetMapping
    public Result<List<Category>> listCategories(
            @RequestParam(required = false) Long parentId,
            @RequestParam(required = false) Integer level,
            @RequestParam(required = false) Integer status) {
        return Result.success(List.of());
    }
    
    @PostMapping
    public Result<Category> createCategory(@Valid @RequestBody CreateCategoryReq req) {
        return Result.success(new Category());
    }
    
    @GetMapping("/{id}")
    public Result<Category> getCategory(@PathVariable Long id) {
        return Result.success(new Category());
    }
    
    @PutMapping("/{id}")
    public Result<Category> updateCategory(@PathVariable Long id, @Valid @RequestBody UpdateCategoryReq req) {
        return Result.success(new Category());
    }
    
    @DeleteMapping("/{id}")
    public Result<Void> deleteCategory(@PathVariable Long id) {
        return Result.success(null);
    }
    
    @GetMapping("/tree")
    public Result<List<CategoryTreeVO>> getCategoryTree() {
        return Result.success(List.of());
    }
    
    @PutMapping("/{id}/move")
    public Result<Category> moveCategory(@PathVariable Long id, @RequestParam Long parentId) {
        return Result.success(new Category());
    }
    
    @PutMapping("/{id}/sort")
    public Result<Category> updateSort(@PathVariable Long id, @RequestParam Integer sortOrder) {
        return Result.success(new Category());
    }
    
    @PutMapping("/{id}/status")
    public Result<Category> updateStatus(@PathVariable Long id, @RequestParam Integer status) {
        return Result.success(new Category());
    }
    
    @GetMapping("/{id}/products")
    public Result<PageResult<Object>> getCategoryProducts(
            @PathVariable Long id,
            @RequestParam(defaultValue = "1") Integer pageNum,
            @RequestParam(defaultValue = "10") Integer pageSize) {
        return Result.success(new PageResult<>());
    }
    
    @GetMapping("/{id}/children")
    public Result<List<Category>> getChildren(@PathVariable Long id) {
        return Result.success(List.of());
    }
    
    @GetMapping("/{id}/parent")
    public Result<Category> getParent(@PathVariable Long id) {
        return Result.success(new Category());
    }
    
    @GetMapping("/{id}/path")
    public Result<List<Category>> getCategoryPath(@PathVariable Long id) {
        return Result.success(List.of());
    }
    
    @PostMapping("/batch")
    public Result<List<Category>> batchCreate(@Valid @RequestBody List<CreateCategoryReq> reqs) {
        return Result.success(List.of());
    }
    
    @DeleteMapping("/batch")
    public Result<Void> batchDelete(@RequestBody List<Long> ids) {
        return Result.success(null);
    }
    
    @GetMapping("/popular")
    public Result<List<Category>> getPopularCategories(@RequestParam(defaultValue = "10") Integer limit) {
        return Result.success(List.of());
    }
}

class CreateCategoryReq {
    private String name;
    private Long parentId;
    private Integer level;
    private Integer sortOrder;
    private String icon;
    
    public String getName() { return name; }
    public void setName(String name) { this.name = name; }
    public Long getParentId() { return parentId; }
    public void setParentId(Long parentId) { this.parentId = parentId; }
    public Integer getLevel() { return level; }
    public void setLevel(Integer level) { this.level = level; }
    public Integer getSortOrder() { return sortOrder; }
    public void setSortOrder(Integer sortOrder) { this.sortOrder = sortOrder; }
    public String getIcon() { return icon; }
    public void setIcon(String icon) { this.icon = icon; }
}

class UpdateCategoryReq {
    private String name;
    private Integer sortOrder;
    private String icon;
    
    public String getName() { return name; }
    public void setName(String name) { this.name = name; }
    public Integer getSortOrder() { return sortOrder; }
    public void setSortOrder(Integer sortOrder) { this.sortOrder = sortOrder; }
    public String getIcon() { return icon; }
    public void setIcon(String icon) { this.icon = icon; }
}

class CategoryTreeVO {
    private Long id;
    private String name;
    private Long parentId;
    private Integer level;
    private String icon;
    private List<CategoryTreeVO> children;
    
    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }
    public String getName() { return name; }
    public void setName(String name) { this.name = name; }
    public Long getParentId() { return parentId; }
    public void setParentId(Long parentId) { this.parentId = parentId; }
    public Integer getLevel() { return level; }
    public void setLevel(Integer level) { this.level = level; }
    public String getIcon() { return icon; }
    public void setIcon(String icon) { this.icon = icon; }
    public List<CategoryTreeVO> getChildren() { return children; }
    public void setChildren(List<CategoryTreeVO> children) { this.children = children; }
}

File: en/kotlin-springmvc/src/main/kotlin/com/example/controller/CategoryController.kt
package com.example.controller

import com.example.dto.PageResult
import com.example.dto.Result
import com.example.model.Category
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/categories")
class CategoryController {
    
    @GetMapping
    fun listCategories(
        @RequestParam(required = false) parentId: Long?,
        @RequestParam(required = false) level: Int?,
        @RequestParam(required = false) status: Int?
    ): Result<List<Category>> {
        val categories = listOf<Category>()
        return Result.success(categories)
    }
    
    @PostMapping
    fun createCategory(@RequestBody req: CreateCategoryReq): Result<Category> {
        val category = Category()
        return Result.success(category)
    }
    
    @GetMapping("/{id}")
    fun getCategory(@PathVariable id: Long): Result<Category> {
        val category = Category()
        return Result.success(category)
    }
    
    @PutMapping("/{id}")
    fun updateCategory(@PathVariable id: Long, @RequestBody req: UpdateCategoryReq): Result<Category> {
        val category = Category()
        return Result.success(category)
    }
    
    @DeleteMapping("/{id}")
    fun deleteCategory(@PathVariable id: Long): Result<Void> {
        return Result.success(null)
    }
    
    @GetMapping("/tree")
    fun getCategoryTree(): Result<List<CategoryTreeVO>> {
        val tree = listOf<CategoryTreeVO>()
        return Result.success(tree)
    }
    
    @PutMapping("/{id}/move")
    fun moveCategory(@PathVariable id: Long, @RequestParam parentId: Long): Result<Category> {
        val category = Category()
        return Result.success(category)
    }
    
    @PutMapping("/{id}/sort")
    fun updateSort(@PathVariable id: Long, @RequestParam sortOrder: Int): Result<Category> {
        val category = Category()
        return Result.success(category)
    }
    
    @PutMapping("/{id}/status")
    fun updateStatus(@PathVariable id: Long, @RequestParam status: Int): Result<Category> {
        val category = Category()
        return Result.success(category)
    }
    
    @GetMapping("/{id}/products")
    fun getCategoryProducts(
        @PathVariable id: Long,
        @RequestParam(defaultValue = "1") pageNum: Int,
        @RequestParam(defaultValue = "10") pageSize: Int
    ): Result<PageResult<Any>> {
        val result = PageResult<Any>()
        return Result.success(result)
    }
    
    @GetMapping("/{id}/children")
    fun getChildren(@PathVariable id: Long): Result<List<Category>> {
        val children = listOf<Category>()
        return Result.success(children)
    }
    
    @GetMapping("/{id}/parent")
    fun getParent(@PathVariable id: Long): Result<Category> {
        val parent = Category()
        return Result.success(parent)
    }
    
    @GetMapping("/{id}/path")
    fun getCategoryPath(@PathVariable id: Long): Result<List<Category>> {
        val path = listOf<Category>()
        return Result.success(path)
    }
    
    @PostMapping("/batch")
    fun batchCreate(@RequestBody reqs: List<CreateCategoryReq>): Result<List<Category>> {
        val categories = listOf<Category>()
        return Result.success(categories)
    }
    
    @DeleteMapping("/batch")
    fun batchDelete(@RequestBody ids: List<Long>): Result<Void> {
        return Result.success(null)
    }
    
    @GetMapping("/popular")
    fun getPopularCategories(@RequestParam(defaultValue = "10") limit: Int): Result<List<Category>> {
        val categories = listOf<Category>()
        return Result.success(categories)
    }
}

data class CreateCategoryReq(
    val name: String?,
    val parentId: Long?,
    val level: Int?,
    val sortOrder: Int?,
    val icon: String?
)

data class UpdateCategoryReq(
    val name: String?,
    val sortOrder: Int?,
    val icon: String?
)

data class CategoryTreeVO(
    val id: Long? = null,
    val name: String? = null,
    val parentId: Long? = null,
    val level: Int? = null,
    val icon: String? = null,
    val children: List<CategoryTreeVO>? = null
)
